e2e/src/test/java/org/jdbi/v3/e2e/TestIssue2623.java

```diff
@@ -0,0 +1,120 @@
+ * limitations under the License.
+ */
+package org.jdbi.v3.e2e;
+
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.util.Objects;
+
+import com.google.common.collect.ImmutableMultimap;
+import com.google.common.collect.Multimap;
+import org.jdbi.v3.core.generic.GenericType;
+import org.jdbi.v3.core.mapper.RowMapper;
+import org.jdbi.v3.core.statement.StatementContext;
+import org.jdbi.v3.guava.GuavaPlugin;
+import org.jdbi.v3.sqlobject.SqlObjectPlugin;
+import org.jdbi.v3.sqlobject.config.KeyColumn;
+import org.jdbi.v3.sqlobject.config.RegisterConstructorMapper;
+import org.jdbi.v3.sqlobject.statement.SqlQuery;
+import org.jdbi.v3.testing.junit5.JdbiExtension;
+import org.jdbi.v3.testing.junit5.internal.TestingInitializers;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.RegisterExtension;
+
+import static org.assertj.guava.api.Assertions.assertThat;
+
+public class TestIssue2623 {
+
+    @RegisterExtension
+    public JdbiExtension h2Extension = JdbiExtension.h2()
+        .withInitializer(TestingInitializers.usersWithData())
+        .withPlugins(new SqlObjectPlugin(), new GuavaPlugin());
+
+    @BeforeEach
+    public void setUp() {
+        h2Extension.getJdbi().registerRowMapper(new GenericType<>() {}, new UserIntegerMapper());
+    }
+
+    @Test
+    public void testParameterizeTypedMapper() {
+        var jdbi = h2Extension.getJdbi();
+
+        Multimap<Integer, User<Integer>> result = jdbi.withExtension(Dao.class, dao -> dao.getUsers(1));
+        assertThat(result).isNotNull().containsAllEntriesOf(ImmutableMultimap.of(1, new User<>("Alice", 1)));
+    }
+
+    @Test
+    public void testConcreteTypedMapper() {
+        var jdbi = h2Extension.getJdbi();
+
+        Multimap<Integer, IntUser> result = jdbi.withExtension(Dao.class, dao -> dao.getIntUsers(1));
+        assertThat(result).isNotNull().containsAllEntriesOf(ImmutableMultimap.of(1, new IntUser("Alice", 1)));
+    }
+
+    public interface Dao {
+        @SqlQuery("select name, id from users where id = :id")
+        @KeyColumn("id")
+        Multimap<Integer, User<Integer>> getUsers(int id);
+
+        @SqlQuery("select name, id from users where id = :id")
+        @KeyColumn("id")
+        @RegisterConstructorMapper(IntUser.class)
+        Multimap<Integer, IntUser> getIntUsers(int id);
+    }
+
+    public record User<T>(String name, T id) {}
+
+    public static class UserIntegerMapper implements RowMapper<User<Integer>> {
+        @Override
+        public User<Integer> map(ResultSet rs, StatementContext ctx) throws SQLException {
+            return new User<>(rs.getString("name"), rs.getInt("id"));
+        }
+    }
+
+    public static class IntUser {
+        private final String name;
+        private final Integer id;
+
+        public IntUser(String name, Integer id) {
+            this.name = name;
+            this.id = id;
+        }
+
+        public String getName() {
```

Comment:
This could be a `record` too to save some boilerplate hashCode / equals?

Reply:
again, same goal. Show the actual implementation. Correct, it could be a record, too.